File: Data Structures and Algorithms/4. Recursion/Kunal/Arrays/BinarySearch.java
import java.util.*;
public class BinarySearch{
    public static int binary_search(int[] arr, int target, int start, int end) {
        if(start > end) return -1;
        int mid = start + (end - start) / 2;

        if(arr[mid] == target)
            return mid;
        else if(arr[mid] > target)
            return binary_search(arr, target, start, mid - 1);
        else
            return binary_search(arr, target, mid + 1, end);
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int []arr = new int[n];
        for(int i = 0; i < n; i++){
            arr[i] = sc.nextInt();
        }
        int target = sc.nextInt();
        int result = binary_search(arr, target, 0, arr.length - 1);
        System.out.printf("%d found at index: %d", target, result);
    }
}

File: Data Structures and Algorithms/4. Recursion/Kunal/Arrays/BinarySearch.cpp
#include <bits/stdc++.h>
using namespace std;

int binary_search(vector<int>& a, int target, int start, int end){

    if(start > end)
        return -1;
    int mid = (start + (end - start)) / 2;
    if(a[mid] == target)
        return mid;
    else if(target < a[mid])
        return (a, target, start, mid - 1);
    else
        return (a, target, mid + 1, end);
}


int main(){
    int n;	cin >> n;
    vector<int> a(n);
    for(int i = 0; i < n; i++){
        cin >> a[i];
    }
    int target;
    cin >> target;
    int result = binary_search(a, target, 0, n - 1);
    cout << "Target found at index: " << result << '\n';
    return 0;
}
